### PR TITLE
Coordinator refactoring

### PR DESCRIPTION
**Pull request by little-dude**
Each commit can be reviewed separately. The main change is 3ea765f.

I'd like to _maybe_ refactor further at some point, taking inspiration from this blog post https://hoverbear.org/blog/rust-state-machine-pattern/

### CODE REVIEW

**Comment by little-dude** [comment]
> minor request: can you also see if your refactoring applies to `Participant.rs` to include as part of this MR? e.g. I've seen some `url` / `hash` occurrences there.

Good point, will do

File: rust/src/coordinator.rs [hunk 1]
```diff
@@ -230,31 +227,24 @@ impl Coordinator {
...
-        .ok_or(PetError::InsufficientParticipants)
-    }
-
     /// Update the mask dictionary.
-    fn update_dict_mask(&mut self, mask_url: &[u8]) {
-        self.dict_mask += mask_url
-            .chunks_exact(mask_url.len())
-            .map(|mask| mask.to_vec());
+    fn update_dict_mask(&mut self, mask_hash: &[u8]) {
+        self.dict_mask.update(iter::once(mask_hash.to_vec()))
```
Comment (janpetschexain) [inline rust/src/coordinator.rs:232]: what's the reason for the change from `+=` to `.update()`? they are equivalent for `Counter<_>`, but the `+=` better conveys the semantics of counting, imo.

File: rust/src/coordinator.rs [hunk 2]
```diff
@@ -230,31 +227,24 @@ impl Coordinator {
...
     fn freeze_dict_mask(&self) -> Result<Vec<u8>, PetError> {
         let counts = self.dict_mask.most_common();
-        (counts.iter().map(|(_, count)| count).sum::<usize>() >= self.min_sum
-            && (counts.len() == 1 || counts[0].1 > counts[1].1))
-            .then(|| counts[0].0.clone())
-            .ok_or(PetError::InsufficientParticipants)
+
+        // it is an error to call freeze_dict_mask if we didn't
+        // receive a single mask
+        assert!(!counts.is_empty());
```
Comment (janpetschexain) [inline rust/src/coordinator.rs:241]: this is already covered by `has_enough_masks()`, the `mask_count` yields `0` in this case.

File: rust/src/coordinator.rs [hunk 3]
```diff
@@ -317,32 +317,60 @@ impl Coordinator {
         self.phase = Phase::Sum;
     }
 
-    /// End the sum phase and proceed to the update phase. Fails due to insufficient sum
-    /// participants.
-    fn proceed_phase_update(&mut self) -> Result<(), PetError> {
-        self.freeze_dict_sum()?;
+    /// Check whether there are enough sum participants for starting
+    /// the update phase.
+    fn has_enough_sum_participants(&self) -> bool {
```
Comment (janpetschexain) [inline rust/src/coordinator.rs:322]: i would suggest a more consistent naming for those three methods, like
```
has_enough_sums()
has_enough_updates()
has_enough_masks()
```
or alternatively
```
has_enough_sum_messages()
has_enough_update_messages()
has_enough_sum2_messages()
```

File: rust/src/lib.rs [hunk 4]
```diff
@@ -20,4 +20,5 @@ pub enum PetError {
     InsufficientSystemEntropy,
     InvalidMessage,
     InsufficientParticipants,
+    MaskSelectionFailed,
```
Comment (janpetschexain) [inline rust/src/lib.rs:23]: maybe `AmbiguousMasks` would fit better here

File: rust/src/service.rs [hunk 5]
```diff
@@ -60,10 +60,10 @@ impl Service {
     //
     // when we fail to handle a message, we should send back and error
     // so that the participant is informed. One problem currently is
-    // that the `validate` methods don't return specific errors.
+    // that the `handle` methods don't return specific errors.
```
Comment (janpetschexain) [inline rust/src/service.rs:63]: on an important side note: the coordinator **must never** inform a participant about the specific reason why message validation failed. that would open gates to various attacks. instead the coordinator must ignore that message, but it can/should log the failure internally.
it is ok to, for example, send an ACK to the participant if the message was received. it might also send a generic failure message if validation fails, but there is no reason to do so and it's better to not open up to potential attacks.

File: rust/src/coordinator.rs (hunk 1 above)
Comment (little-dude) [reply rust/src/coordinator.rs:232]: Having a `+` operator with operands of that of different types was very confusing, especially with types that usually don't work with `+`: `Dict<K,V> + `Iterator<Item=V>` doesn't make much sense I think.

File: rust/src/coordinator.rs (hunk 2 above)
Comment (little-dude) [reply rust/src/coordinator.rs:241]: Yeah that's why I made is an `assert!`. But you're right, maybe it's useless I'll remove it.

File: rust/src/coordinator.rs (hunk 3 above)
Comment (little-dude) [reply rust/src/coordinator.rs:322]: ah very nice. I have a slight preference the first version. How about you (or @finiteprods) ?

File: rust/src/lib.rs (hunk 4 above)
Comment (little-dude) [reply rust/src/lib.rs:23]: Thanks!

File: rust/src/coordinator.rs (hunk 3 above)
Comment (finiteprods) [reply rust/src/coordinator.rs]: same here

Comment (janpetschexain) [reply rust/src/coordinator.rs]: same

Comment (little-dude) [reply rust/src/coordinator.rs]: alright it's settled then.

File: rust/src/coordinator.rs (hunk 1 above)
Comment (janpetschexain) [reply rust/src/coordinator.rs:232]: mhh ok, that's probably quite subjective. i'm ok going with `.update()`.
